src/psyclone/dynamo0p3.py

```diff
@@ -7452,137 +7451,136 @@ def gen_code(self, parent):
+        if actual_n_args != expected_n_args:
+            raise GenerationError(
+                "The number of arguments expected by the psy-layer kernel "
+                "call '{0}' does not match the actual number of kernel "
+                "arguments '{1}' in kernel '{2}'.".format(
+                    expected_n_args, actual_n_args, self.name))
+
+        # 2: Check that the properties of each argument matches
+        for idx, kern_code_arg in enumerate(kern_code_args):
+            interface_arg = interface_args[idx]
+            self._validate_kernel_code_arg(kern_code_arg, interface_arg)
+
+    def _validate_kernel_code_arg(self, kern_code_arg, interface_arg):
+        '''Internal method to check that the supplied datatypes match and
+        raise appropriate exceptions if not.
+
+        :param kern_code_arg: kernel code argument.
+        :type kern_code_arg: :py:class:`psyclone.psyir.symbols.DataSymbol`
+        :param interface_arg: expected argument.
+        :type interface_arg: :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+        :raises GenerationError: if the contents of the arguments do \
+            not match.
+        :raises InternalError: if an unexpected datatype is found.
+
+        '''
+        # 1: intrinsic datatype
+        actual_datatype = kern_code_arg.datatype.intrinsic
+        expected_datatype = interface_arg.datatype.intrinsic
+        if actual_datatype != expected_datatype:
+            raise GenerationError(
+                "Kernel argument '{0}' has datatype '{1}' "
+                "in kernel '{2}' but the LFRic API expects '{3}'."
+                "".format(kern_code_arg.name, actual_datatype,
+                          self.name, expected_datatype))
+        # 2: precision
+        actual_precision = kern_code_arg.datatype.precision
+        expected_precision = interface_arg.datatype.precision
+        if actual_precision.name != expected_precision.name:
+            raise GenerationError(
+                "Kernel argument '{0}' has precision '{1}' "
+                "in kernel '{2}' but the LFRic API expects '{3}'."
+                "".format(kern_code_arg.name, actual_precision.name,
+                          self.name, expected_precision.name))
+        # 3: intent
+        actual_intent = kern_code_arg.interface.access
+        expected_intent = interface_arg.interface.access
+        if actual_intent.name != expected_intent.name:
+            raise GenerationError(
+                "Kernel argument '{0}' has intent '{1}' "
+                "in kernel '{2}' but the LFRic API expects intent '{3}'."
+                "".format(kern_code_arg.name, actual_intent.name,
+                          self.name, expected_intent.name))
+        # 4 scalar or array
+        if interface_arg.is_scalar:
+            if not kern_code_arg.is_scalar:
+                raise GenerationError(
+                    "Kernel argument '{0}' is expected to be a scalar "
```

Comment:
"Argument '{0}' to kernel '{1}' should be a scalar according to the LFRic API but it is not."

Reply:
Much nicer thanks.